### PR TITLE
Update interstitial class

### PR DESCRIPTION
**Pull request by samwaseda**
Following the spontaneous meeting between @pmrv and me, I updated the interstitial class, in the hope that it does what it's supposed to do now.

Main changes:

- Now the workflow is more easily accessible
- Classification takes place following the Steinhardt parameters

### CODE REVIEW

**Comment by jan-janssen** [comment]
@samwaseda Can you add tests for this functionality? To me it seems the functionality changed but all the tests pass, so maybe it is not sufficiently tested.

**Comment by samwaseda** [comment]
It tests the full functionality instead of individual functions, which is the reason why it doesn't fail, but ok I can also add individual tests.

File: structuretoolkit/analyse/spatial.py [hunk 1]
```diff
@@ -55,22 +55,69 @@ def get_mean_positions(positions, cell, pbc, labels):
...
+        positions = get_wrapped_coordinates(structure=structure, positions=positions)
+        unique_indices = np.unique(
+            np.round(positions, decimals=decimals), axis=0, return_index=True
+        )[1]
+        positions = positions[unique_indices]
+    raise ValueError("High symmetry points could not be detected")
+
+
+def cluster_by_steinhardt(positions, neigh, l_values, q_eps, var_ratio, min_samples):
+    from sklearn.cluster import DBSCAN
```
Comment (pmrv) [inline structuretoolkit/analyse/spatial.py:89]:
```suggestion
    """
    Clusters candidate positions via Steinhardt parameters and the variance in distances to host atoms.
    
    The cluster that has the lowest variance is returned, i.e. those positions that have the most "regular" coordination polyhedra.
    
    Args:
        positions (array): candidate positions
        neigh (Neighbor): neighborhood information of the candidate positions
        l_values (list of int): which steinhardt parameters to use for clustering
        q_eps (float): maximum intercluster distance in steinhardt parameters for DBSCAN clustering
        var_ratio (float): multiplier to make steinhardt's and distance variance numerically comparable
        min_samples (int): minimum size of clusters
    
    Returns:
         array:  Positions of the most likely interstitial sites
    """
    from sklearn.cluster import DBSCAN
```

File: structuretoolkit/analyse/spatial.py [hunk 2]
```diff
@@ -55,22 +55,69 @@ def get_mean_positions(positions, cell, pbc, labels):
...
+
+    if min_samples is None:
+        min_samples = min(len(neigh.distances), 5)
+    neigh = neigh.get_neighborhood(positions)
+    Q_values = np.array([neigh.get_steinhardt_parameter(ll) for ll in l_values])
+    db = DBSCAN(q_eps, min_samples=min_samples)
+    var = np.std(neigh.distances, axis=-1)
+    descriptors = np.concatenate((Q_values, [var * var_ratio]), axis=0)
```
Comment (pmrv) [inline structuretoolkit/analyse/spatial.py:97]: Did you consider `sklearn.preprocessing.StandardScaler`?

File: structuretoolkit/analyse/spatial.py [hunk 3]
```diff
@@ -55,22 +55,69 @@ def get_mean_positions(positions, cell, pbc, labels):
...
+        1. Eliminate points within a distance from the nearest neighboring atoms as
             given by `min_distance`
-        3. Initialize neighbor environment using `get_neighbors`
-        4. Shift interstitial candidates to the nearest symmetric points with respect to the
+        2. Shift interstitial candidates to the nearest symmetric points with respect to the
             neighboring atom sites/vertices.
-        5. Kick out points with large neighbor distance variances; this eliminates "irregular"
-            shaped interstitials
-        6. Cluster interstitial candidates to avoid point overlapping.
+        3. Cluster interstitial candidates to avoid point overlapping.
```
Comment (pmrv) [inline structuretoolkit/analyse/spatial.py:117]:
```suggestion
        3. Cluster interstitial candidate positions to avoid point overlapping.
```

**Review by pmrv — APPROVED** [review]
I left some docstring nits, especially for `cluster_by_steinhardt`, because that's non-obvious for other devs.  Otherwise lgtm.

Tests are imo as they check that known interstitials are found in simple crystals.

**Comment by pmrv** [comment]
Ah also I forget: This removes the iterative "refinement" of interstitial positions, in the last session we never used it anymore, but are there cases where it still might? I think you mentioned bcc and that's covered by the tests so might just be remembering it wrong.

**Comment by samwaseda** [comment]
> Ah also I forget: This removes the iterative "refinement" of interstitial positions, in the last session we never used it anymore, but are there cases where it still might? I think you mentioned bcc and that's covered by the tests so might just be remembering it wrong.

It still does the refinement, but it's not necessary to set it anymore because the algorithm checks the convergence automatically.

**Pull request merged**